lang/c++/impl/Zigzag.cc

```diff
@@ -20,66 +20,39 @@
 #include "Zigzag.hh"
 
 namespace avro {
-
-uint64_t
-encodeZigzag64(int64_t input)
-{
-    // cppcheck-suppress shiftTooManyBitsSigned
-    return ((input << 1) ^ (input >> 63));
-}
-
-int64_t
-decodeZigzag64(uint64_t input)
-{
-    return static_cast<int64_t>(((input >> 1) ^ -(static_cast<int64_t>(input) & 1)));
-}
-
-uint32_t
-encodeZigzag32(int32_t input)
-{
-    // cppcheck-suppress shiftTooManyBitsSigned
-    return ((input << 1) ^ (input >> 31));
-}
-
-int32_t
-decodeZigzag32(uint32_t input)
-{
-    return static_cast<int32_t>(((input >> 1) ^ -(static_cast<int64_t>(input) & 1)));
-}
-
+// TODO: The following two functions have exactly the same code except for the type.
+// They should be implemented as a template.
 size_t
-encodeInt64(int64_t input, std::array<uint8_t, 10> &output)
-{
-    // get the zigzag encoding
-    uint64_t val = encodeZigzag64(input);
+encodeInt64(int64_t input, std::array<uint8_t, 10> &output) noexcept {
+    auto val = encodeZigzag64(input);
 
     // put values in an array of bytes with variable length encoding
-    const int mask  = 0x7F;
-    output[0] = val & mask;
-    size_t bytesOut = 1;
-    while( val >>=7 ) {
-        output[bytesOut-1] |= 0x80;
-        output[bytesOut++] = (val & mask);
+    const int mask = 0x7F;
+    auto v = val & mask;
+    size_t bytesOut = 0;
+    while (val >>= 7) {
+        output[bytesOut++] = (v | 0x80);
+        v = val & mask;
     }
 
+    output[bytesOut++] = v;
     return bytesOut;
 }
 
 size_t
-encodeInt32(int32_t input, std::array<uint8_t, 5> &output)
-{
-    // get the zigzag encoding
-    uint32_t val = encodeZigzag32(input);
+encodeInt32(int32_t input, std::array<uint8_t, 5> &output) noexcept {
+    auto val = encodeZigzag32(input);
 
     // put values in an array of bytes with variable length encoding
-    const int mask  = 0x7F;
-    output[0] = val & mask;
-    size_t bytesOut = 1;
-    while( val >>=7 ) {
-        output[bytesOut-1] |= 0x80;
-        output[bytesOut++] = (val & mask);
+    const int mask = 0x7F;
+    auto v = val & mask;
```

Comment:
it might be better to set the type explicitly here to make it clear this is intended to be unsigned (same note above on auto)

Reply:
sorry for the comments I see this is already merged.